torchrl/envs/vec_env.py

```diff
@@ -1076,3 +1089,257 @@ def _run_worker_pipe_shared_mem(
+        else:
+            ret = torch.tensor(x)
+        return ret
+
+    def _treevalue_to_dict(self, tv: "treevalue.TreeValue") -> Dict[str, Any]:
+        """Converts TreeValue to a dictionary.
+
+        Currently only supports depth 1 trees, but can easily be extended to arbitrary depth if necessary.
+        """
+        return {k[0]: torch.tensor(v) for k, v in treevalue.flatten(tv)}
+
+    def _set_seed(self, seed: Optional[int]):
+        if seed is not None:
+            print(
+                "MultiThreadedEnvWrapper._set_seed ignored, as setting seed in an existing envorinment is not\
+                   supported by envpool. Please create a new environment, passing the seed to the constructor."
+            )
+
+
+class MultiThreadedEnv(MultiThreadedEnvWrapper):
+    """Multithreaded execution of environments based on EnvPool.
+
+    An alternative to ParallelEnv based on multithreading. It's faster, as it doesn't require new process spawning, but
+    less flexible, as it only supports environments implemented in EnvPool library.
+    Currently only supports synchronous execution mode, when the batch size is equal to the number of workers, see
+    https://envpool.readthedocs.io/en/latest/content/python_interface.html#batch-size.
+
+    >>> env = MultiThreadedEnv(num_workers=3, env_name="Pendulum-v1")
+    >>> env.reset()
+    >>> env.rand_step()
+    >>> env.rollout(5)
+    >>> env.close()
+
+    Args:
+        num_workers: number of worker threads to create.
+        env_name: name of the environment, corresponding to task_id in EnvPool.
+        create_env_kwargs: additional arguments which will be passed to envpool.make.
+    """
+
+    def __init__(
+        self,
+        num_workers: int,
+        env_name: str,
+        create_env_kwargs: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ):
+
+        self.env_name = env_name.replace("ALE/", "")  # Naming convention of EnvPool
+        self.num_workers = num_workers
+        self.batch_size = torch.Size([num_workers])
+        self.create_env_kwargs = create_env_kwargs or {}
+
+        kwargs["num_workers"] = num_workers
+        kwargs["env_name"] = self.env_name
+        kwargs["create_env_kwargs"] = create_env_kwargs
+        super().__init__(**kwargs)
+
+    def _build_env(
+        self,
+        env_name: str,
+        num_workers: int,
+        create_env_kwargs: Optional[Dict[str, Any]],
+    ) -> Any:
+        create_env_kwargs = create_env_kwargs or {}
+        env = envpool.make(
+            task_id=env_name,
+            env_type="gym",
+            num_envs=num_workers,
+            gym_reset_return_info=True,
+            **create_env_kwargs,
+        )
+        return super()._build_env(env)
+
+    def _set_seed(self, seed: Optional[int]):
+        """Library EnvPool only support setting a seed by recreating the environment."""
+        if seed is not None:
+            self.create_env_kwargs["seed"] = seed
+            self._env = self._build_env(
```

Comment:
I am curious if this is an expensive operation and if we should highlight this behavior via logging (debug level)

Reply:
Added a `logging.debug` message